diff --git a/src/gost_grasshopper_math.h b/src/gost_grasshopper_math.h
index 176d16a..3291267 100644
--- a/src/gost_grasshopper_math.h
+++ b/src/gost_grasshopper_math.h
@@ -118,6 +118,10 @@ static GRASSHOPPER_INLINE void grasshopper_convert128(grasshopper_w128_t* x, con
 extern uint8_t grasshopper_galois_alpha_to[256];
 extern uint8_t grasshopper_galois_index_of[256];
 
+#ifdef _WIN32
+#    define __builtin_expect(Condition, likely) Condition
+#endif
+
 static GRASSHOPPER_INLINE uint8_t grasshopper_galois_mul(uint8_t x, uint8_t y) {
     if (__builtin_expect(x != 0 && y != 0, 1)) {
         return grasshopper_galois_alpha_to[(grasshopper_galois_index_of[x] + grasshopper_galois_index_of[y]) %
